### PR TITLE
util, memory and math-core package separation

### PR DESCRIPTION
**Pull request by holloszaboakos**
I tried te separate the memory modul from foundation.
I had to leave some features there because of dependency on the Half class from number.
I also had to separate the util package and create tha math-core modul for broadly used math functionality.
I included memory and math-core in datastructures, so I could remove meny internal functionality.

### CODE REVIEW

File: korlibs-math-core/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        "https://raw.githubusercontent.com/korlibs/korge/main/korlibs-math-core/LICENSE"
+    )
+}
+
+dependencies {
+    commonMainApi(libs.kotlinx.coroutines.core)
+    commonMainApi(libs.kotlinx.atomicfu)
+    commonTestApi(libs.kotlinx.coroutines.test)
+    commonMainApi(libs.kotlinx.atomicfu)
+    commonTestApi(libs.kotlinx.coroutines.test)
```
Comment (soywiz) [inline korlibs-math-core/build.gradle.kts:18]: Duplicated lines. Why kotlinx.coroutines is required here?

**Comment by soywiz** [comment]
Thanks @holloszaboakos !

For the sake of being easier to review. Let's split this PR: 
* `inject` in one PR
* `math-core` in other PR
  * After `math-core` is merged, let's update `korlibs-memory` in other PR
  * After `math-core` is merged, let's update `korlibs-util` in other PR

**Comment by holloszaboakos** [comment]
I will, but first comment on the direction pls. Are these modules and their name fine by you? What about the memory code left in foundation?

**Comment by soywiz** [comment]
> I will, but first comment on the direction pls. Are these modules and their name fine by you? What about the memory code left in foundation?

Look reasonable to me. Since we are not going to change change packages or class names, it would be just renaming a folder, and changing a string in 2 o 3 files. So we can change our mind easily if we find better names. But as said, I believe they are reasaonable already.

**Comment by holloszaboakos** [comment]
it will be sliced into multiple PR-s

**Pull request closed without merging**